### PR TITLE
Joint trajectory controller

### PR DESCRIPTION
**Pull request by adolfo-rt**
This is a fully functional and tested implementation of a Joint Trajectory Controller.

**Note:** I developed this controller in my Fuerte backport of ros_controllers, and tried to make this PR Hydro-friendly, but I've not tested it in a catkin system. I suggest that this PR should be merged when someone using Groovy or Hydro is willing to iron out any pending wrinkles resulting from the Hydro port. It should suffice to run the test suite to ensure that everything is working as expected.
### ROS API
- Exposes two ROS APIs for sending commands `FollowJointTrajectory` action and `trajectory_msgs::JointTrajectory` topic.
- Current controller state is available in a published `control_msgs::JointTrajectoryControllerState` topic.
- Controller state at any future time can be queried through a service of type `control_msgs/JointTrajectoryControllerState`.
### Segment type
- Controller is templated on the segment type, so any segment implementation conforming to the expected interface can be used.
- Controllers for new segment types are a template specialization away.
- Multi-dof quintic spline segment implementation provided by default.
### Hardware interface type
- Controller is templated on the hardware interface type.
- Any hardware interface can be supported as long as a template specialization for the `HardwareInterfaceAdapter` exists.
- Controller currently can work with position and effort (untested) hardware interfaces.
### Other
- Realtime-safe.
- Proper handling of wrapping (continuous) joints.
- Discontinuous system clock changes do not cause discontinuities in the execution of already queued trajectory segments.
- Complete test suite and developer documentation.
### TODO
- User-oriented documentation.

### CODE REVIEW

**Comment by jbohren** [comment]
Cool stuff, I'll look through this closely tomorrow. How hard do you think it would be to split the majority of the controller into a framework-independent library?

**Comment by adolfo-rt** [comment]
The trajectory representation and sampling code is already in separate ros-agnostic headers. This accounts for all the core computations.

The code inside the actual controller is mostly for managing the ROS API, and just a little bit of it is ros_control-specific. It still amounts to some ~500 sloc.

Some extra work would be required to say, implement the ROS API in Orocos RTT, and using Orocos ports, lock-free buffers, the native actionlib server you've been working on, etc. It should be doable though, in the end we're talking mostly about methods for init, start, update and stop, plus callbacks.

**Comment by davetcoleman** [comment]
I'll try to compile this now on Hydro...

**Comment by davetcoleman** [comment]
For others - note that https://github.com/ros-controls/realtime_tools/pull/4 is required to build this

File: joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+    goal_handle_timer_ = controller_nh_.createTimer(action_monitor_period_,
+                                                    &RealtimeGoalHandle::runNonRealtime,
+                                                    rt_goal);
+    goal_handle_timer_.start();
+  }
+  else
+  {
+    // Reject goal. Determine if the reason was invalid joints or something else
+    using internal::permutation;
+    std::vector<unsigned int> permutation_vector = permutation(joint_names_, gh.getGoal()->trajectory.joint_names);
```
Comment (davetcoleman) [inline joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:527]: Looks like you have a miss match of using a long unsigned int and just an unsigned int, but adding long caused other problems:

```
/home/dave/ros/ws_gazebo/src/ros_controllers/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:318:3:   instantiated from ‘bool joint_trajectory_controller::JointTrajectoryController<SegmentImpl, HardwareInterface>::init(hardware_interface::PositionJointInterface*, ro\
s::NodeHandle&, ros::NodeHandle&) [with SegmentImpl = trajectory_interface::QuinticSplineSegment<double>, HardwareInterface = hardware_interface::PositionJointInterface]’
/home/dave/ros/ws_gazebo/src/ros_controllers/joint_trajectory_controller/src/joint_trajectory_controller.cpp:59:1:   instantiated from here
/home/dave/ros/ws_gazebo/src/ros_controllers/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:527:114: error: conversion from ‘std::vector<long unsigned int>’ to non-scalar type ‘std::vector<unsigned int>’ requested
```

Comment (davetcoleman) [inline joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h:527]: I tried to fix this myself for a while but couldn't figure out why it thinks the output is of type "long unsigned int"

**Comment by adolfo-rt** [comment]
@davetcoleman please try again, there's a new commit. Sorry for the inconvenience. This was most probably a 32-64 bit issue, where integral size types differ between architectures. The ultimate cause was me being lazy and not enforcing the correct return types.

As a sidenote, I can't wait until good C++11 support reaches stable Ubuntu distros (and ROS), to start coding [AAA-style](http://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto) and make issues like this one something of the past!

**Comment by davetcoleman** [comment]
I reviewed your fixes and tried to replicate them for this new set of similar errors I have, but I'm finding your code a little over-complex with all its typedefs, impl files and structure. Also, I'm just not putting much time into grasping it all. But here are my new errors after pulling in your fixes a few hours ago:

https://gist.github.com/davetcoleman/6309167

**Comment by adolfo-rt** [comment]
I installed gcc-multilib to run both 32 and 64 builds. I could reproduce the error, and should hopefully be fixed now. Hopefully this will land in the official Hydro release.

It's true that the code has lots of typedefs and typenames, as happens with all templated code. Speaking of the future, if we can manage to implement the controller without it being templated, some of the cruft will go away. This means 1. puligin-enabled segments, and 2. an easy way to connect to different hardware interfaces (could be a plugin as well).

@davetcoleman Thanks for the time you're putting into this.

**Comment by davetcoleman** [comment]
It builds, well done! I also ran your tests (wow - you have a lot of good tests) and I got mixed results. Some of the tests failed and some passed, but even the passed ones have errors?

Results from Hydro Precise test run:
https://gist.github.com/davetcoleman/6324130

I'd say we can go ahead and merge this though.

@adolfo-rt today is the last day of my internship at OSRF so I won't be around as much - need to work on papers :)

File: joint_trajectory_controller/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      urdf
+      control_toolbox
+      controller_interface
+      hardware_interface
+      realtime_tools
+      control_msgs
+      trajectory_msgs
+      rostest
+      controller_manager
+      xacro
```
Comment (davetcoleman) [inline joint_trajectory_controller/CMakeLists.txt:65]: I've been having issues depending on xacro in groovy... doesn't seem to work. Probably not a direct issue here but I have to remove these references to compile. What do you think @wjwwood ?

**Comment by adolfo-rt** [comment]
All failing tests are rostests. It seems to be a setup issue. Apparently not all nodes from the .test file are being launched, eg. [this line](https://gist.github.com/davetcoleman/6324130#file-joint_trajectory_controller_test-L158) is failing on a precondition that indicates that the controller to test is not running (or maybe taking too long to start). I'll see if I can sort this out on my end.

Don't worry about passing tests containing ERROR log statements. It's expected behavior, as the test is exercising that it can gracefully handle invalid/inconsistent input.

**Comment by adolfo-rt** [comment]
I setup a Groovy installation and verified that the xacro dependency causes problems, so I commented it for now. Other than that, I still have to figure out how to setup rostests, which are not working. Let's see if I can fix it on time.

**Comment by adolfo-rt** [comment]
I just tested the latest changes in a hydro build, and all tests are passing. I'm giving this a go.

**Comment by davetcoleman** [comment]
Well done!

**Comment by jbohren** [comment]
@adolfo-rt Can you give some more details on the trajectory reversal behavior that you mentioned in 129ea1b?

**Comment by adolfo-rt** [comment]
If a joint currently has a state (pos_1, vel_1), where the velocity vel_1 is nonzero, and you request the joint to stop by sending the command (pos_1, 0) in time t, what will happen is that the joint will move past pos_1 reverse its direction of motion, and then go back and reach pos_1, this time with zero velocity. I believe this is what you had in your original Reflexxes based implementation.

What 129ea1b does is to compute a stop trajectory starting at state (pos_1, vel_1) and with duration t that will stop past pos_1, but without direction reversals. In particular, it generates a segment going from (pos_1, vel_1) to (pos_1, -vel_1) with duration 2t, and then samples it at the midpoint to get pos_2, where (assuming symmetry) the velocity is zero. The stop trajectory then goes from (pos_1, vel_1) to (pos_2, 0) in time t.

The limitation of this approach is that it assumes symmetry in the segment from (pos_1, vel_1) to (pos_1, -vel_1). It is the case for the current spline representation, but might not be the case for other representations, e.g., a Reflexxes based implementation might yield non-symmetric segments when joint limit violations are involved.

Notice that by default the stop trajectory duration is zero, and the deceleration profile is determined by either low-level joint limits enforcing (outside the controller) or by the hardware itself.

**Comment by jbohren** [comment]
> that will stop _past_ pos_1, but without direction reversals

Ah alright. That's what I was wondering about.

> Notice that by default the stop trajectory duration is zero, and the deceleration profile is determined by either low-level joint limits enforcing (outside the controller) or by the hardware itself.

I did notice this. Have you put any time into creating a segment type that incorporates joint velocity limits?

**Comment by adolfo-rt** [comment]
> > Notice that by default the stop trajectory duration is zero, and the deceleration profile is determined by either low-level joint limits enforcing (outside the controller) or by the hardware itself.
> 
> I did notice this. Have you put any time into creating a segment type that incorporates joint velocity limits? 

No, I ran out of time budget when coding the controller. I was pondering whether it made more sense to put it here in the controller or defer limits enforcing to the robot hardware abstraction. The upside of having it in the controller is that effort-controlled robots would benefit from it as well.

Wherever it ended up being implemented, I really liked the idea of combining Reflexxes with the  implementation in JointLimitsInterface. Since Reflexxes does not enforce position limits, we could adjust online the velocity limits depending on how far we are from the position limits.

File: joint_trajectory_controller/CMakeLists.txt (hunk 2 above)
Comment (wjwwood) [inline joint_trajectory_controller/CMakeLists.txt:65]: I think if you are having trouble with `xacro` on Groovy you should file an issue against `xacro`, it should work as far as I know.

Comment (wjwwood) [inline joint_trajectory_controller/CMakeLists.txt:65]: Well, also this was 6 months ago :anguished:, so ignore the noise.

**Pull request merged**